src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClone = Object.assign({}, state);
+  const resultArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          stateClone[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateClone[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+        break;
+      default:
+        throw new Error('Incorrect action type');
+    }
+    resultArray.push(Object.assign({}, stateClone));
```

Comment:
there is no need to create the new objects here with `assign` or you can use destructuring
```suggestion
    resultArray.push({ ...stateClone });
```